src/memory_manager.rs

```diff
@@ -229,6 +229,23 @@ pub fn post_alloc<VM: VMBinding>(
 /// * `src`: The modified source object.
 /// * `slot`: The location of the field to be modified.
 /// * `target`: The target for the write operation.
+///
+/// # Deprecated
+///
+/// This form of write barrier has multiple issues.
+///
+/// -   It is only able to write non-null object references into the slot.  But dynamic language
+///     VMs may write non-reference values, such as tagged small integers, special values such as
+///     `null`, `undefined`, `true`, `false`, etc. into a field that previous contains an object
+///     reference.
+/// -   It relies on `slot.store` to write `target` into the slot, but `slot.store` is designed for
+///     forwarding references when an object is moved by GC, and is supposed to preserve tagged
+///     type information, the offset (if it is an interior pointer), etc.  A write barrier is
+///     associated to an assignment operation, which usually updates such information instead.
+///
+/// VM bindings should use `object_reference_write_pre` and `object_reference_write_post` instead.
+/// Those functions leaves the actual write of the slot to the VM binding.
+#[deprecated = "Use `object_reference_write_pre` and `object_reference_write_post` instead"]
```

Comment:
I don't recall we have any discussion about removing the subsuming barrier. If I missed any previous discussion, just let me know.

If I understand you right, there are a few issues with this method right now:
1. `target: ObjectReference` cannot represent all the cases. If MMTk does the write for the binding, `target` needs to be able to represent any value the runtime may hold. `Option<ObjectReference>` is not sufficient either. We may want something like `target: Value`.
2. `slot: VM::VMEdge` cannot handle writing values that are not references. Though `Edge` (we want to rename it to `Slot`) is introduced for updating slots for copying GC, I think it is a valid case of reusing this trait to update slots in write barriers. We could extend the trait to allow it to store any value.

Is there any fundamental reason that we should remove the subsuming barrier instead of trying to correct it?

The reason why `write_pre/post` work is that we currently do not use the `slot` argument, and we do not care the `target` value if it is not a reference (so `Option<ObjectReference>` is sufficient).

Reply:
We don't *remove* the subsuming barrier, but we need to *redesign* it.  That's why I said "*This form of* write barrier has multiple issues", not subsuming barriers in general.  We will replace this `object_reference_write` with another function, maybe with the same name, but different parameters.

There are ways to allow the VM customize the write operation of the subsuming barrier.  One way is letting the caller givt it a callback that actually writes to the field so that it can write the field in whatever way it likes.  The subsuming barrier effectively executes the pre and post barrier between the callback.

Reply:
I changed the comment and emphasized that we would redesign that function and replace the current one.